### ISSUE TITLE
A11y: Dropdowns have accessibility focus issues

### ISSUE THREAD

**Issue by andywolff** [flutter/flutter#40101]
Internal: b/140831334

If you have a dropdown (DropdownButton and DropdownMenuItems) below a Navigator in a MaterialApp, it breaks the screenreader focus behavior.

The problem is that when you activate the dropdown button, focus gets lost instead of going to the selected item in the popup. Sometimes it works properly the first time you open the popup, but not subsequent times. One strange thing I noticed (TalkBack only) was that if you navigate backwards three times in this state, it will successfully focus on the last element, but navigating forwards from this state never focuses on anything. Also, when you select an item and the popup closes, focus gets lost instead of returning to the button.

If the dropdown does not have a Navigator in its ancestry, it doesn't have this problem.

Our app does have a page where the dropdown focus works correctly even though it has a Navigator in the tree above the dropdown. This gives me hope that there may be some kind of workaround for the issue. However, I haven't been able to isolate what makes this different from the repro case below because the app's tree has many layers.

```
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Navigator(
        onGenerateRoute: (_) {
          return MaterialPageRoute(
            builder: (context) {
              return Scaffold(
                body: SafeArea(
                  child: DropdownButton(
                    value: "foo",
                    items: [
                      DropdownMenuItem(value: "foo", child: Text("Foo"),),
                      DropdownMenuItem(value: "bar", child: Text("Bar"),),
                    ],
                    onChanged: (_) {},
                  ),
                ),
              );
            },
          );
        },
      ),
    );
  }
}
```

Tested on both Android and iOS. Issue is present on both.

**Comment by HansMuller** [comment]
CC @gspencergoog @jonahwilliams

**Comment by gspencergoog** [comment]
I assume that when you're referring to focus navigation here, you are referring accessibility focus within TalkBack/VoiceOver, correct, and not to input focus?

**Comment by andywolff** [comment]
Yes, accessibility focus

**Comment by jkurtw** [comment]
Internal: b/140831334

**Comment by jkurtw** [comment]
breaks a critical application flow for screenreader users, but if they skip the affected page, they can achieve the same effect on a different page in the app

**Comment by gspencergoog** [comment]
We have been unable to reproduce it in iOS: does it have different expression on iOS for you?

**Comment by gspencergoog** [comment]
Update: this appears to be a bug in the way we interface with TalkBack on Android.  TalkBack is somehow finding that there is nothing that can be focused, even though it knows about all of the appropriate nodes, and they all exist and appear to be focusable.  Since it works on iOS and not an Android, we're assuming that the problem isn't in the Flutter framework, but in the engine side of things.

**Comment by andywolff** [comment]
re: iOS: I just retried it and it seems I misunderstood the behavior on iOS. I was getting confused because in the affected dropdowns in our app on iOS, instead of focus getting lost, it just sometimes focuses on the dismiss shield (but you can still easily navigate to the popup items). The focus highlight outline also jumps to the top-left corner briefly, which was throwing me off. The focus going to the dismiss shield doesn't happen in the repro case in the original post, I think the cause is probably unrelated. If I can narrow down a good repro case for this, I will post a separate issue.

However, iOS does appear to have a similar problem when you select an item. It returns focus to the first item on the page instead of the dropdown button. The repro case from the original post is insufficient to show this because the dropdown is the first item on the page, but if you throw it in a Column and add a Text() above it, you can see it. When you select an item, focus goes to "some element" instead of the dropdown button. I haven't checked whether this is also related to the Navigator in the ancestry or just always present.

```
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Navigator(
        onGenerateRoute: (_) {
          return MaterialPageRoute(
            builder: (context) {
              return Scaffold(
                body: SafeArea(
                  child: Column(
                    children: <Widget>[
                      Text("some element"),
                      DropdownButton(
                        value: "foo",
                        items: [
                          DropdownMenuItem(
                            value: "foo",
                            child: Text("Foo"),
                          ),
                          DropdownMenuItem(
                            value: "bar",
                            child: Text("Bar"),
                          ),
                        ],
                        onChanged: (_) {},
                      ),
                    ],
                  ),
                ),
              );
            },
          );
        },
      ),
    );
  }
}
```

**Comment by gspencergoog** [comment]
BTW, I am able to reproduce this on Android without a Navigator in the tree:

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String value = 'foo';

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        body: SafeArea(
          child: Center(
            child: Row(
              mainAxisAlignment: MainAxisAlignment.center,
              mainAxisSize: MainAxisSize.min,
              children: <Widget>[
                DropdownButton(
                  value: value,
                  items: [
                    DropdownMenuItem(
                      value: 'foo',
                      child: Text('Foo'),
                    ),
                    DropdownMenuItem(
                      value: 'bar',
                      child: Text('Bar'),
                    ),
                    DropdownMenuItem(
                      value: 'baz',
                      child: Text('Baz'),
                    ),
                  ],
                  onChanged: (String value) {
                    setState(() {
                      this.value = value;
                    });
                  },
                ),
                MaterialButton(
                  child: Text('Reset'),
                  onPressed: () {
                    setState(() {
                      this.value = 'foo';
                    });
                  },
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
```

**Comment by Norbert515** [comment]
This issue also affects shortcuts on desktop. The new `Shortcuts` widget relies on focus. When pressing a `DropDownButton` a new route is pushed and the focus to the `Shortcuts` is lost, thus shortcuts do not work anymore. 
Looking forward to the PR

**Comment by gspencergoog** [comment]
Yes, that's also a know issue, but that's not related to this one, since that is input focus, not accessibility focus. In that case, it's just because we're not done enabling all widgets with focus handling.

**Comment by gspencergoog** [comment]
OK, I think I've got a fix for this problem.  It will land in two parts: one an engine fix: flutter/engine#12410, and another that changes the behavior for dropdown widgets and popups (and other widgets) to set the input focus bit in the semantics of components that have the "primary focus", and also add regular (non-a11y) focus support to dropdowns and popups.

**Comment by Hixie** [comment]
update: being worked on; there was a performance regression that we're tracking down before we can commit this. Thank you for your patience.

**Comment by keobrien** [comment]
I'm still seeing these issues on iOS TalkBack & Android TalkBack for the DropdownButtonFormField widget.

**Issues:**
1. VoiceOver & TalkBack - DropdownButtonFormField does not gain accessibility focus when using the native keyboard next button from another field. 
1. VoiceOver & TalkBack - After making a selection from DropdownButtonFormField, accessibility focus does not return to the field, accessibility focus is sent to the top of the page.
1. VoiceOver - When opening the DropdownButtonFormField, accessibility focus is sent to the hidden "Dismiss" link at the bottom. A user has to swipe back to get to the dropdown items, accessibility focus should be placed at the top of the list.

**Code Example:**
`
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  FocusNode dropdownFocus = FocusNode();
  String dropdownValue;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Home'),
      ),
      body: GestureDetector(
        child: Column(
          children: [
            TextFormField(
              decoration: InputDecoration(labelText: 'Test Field 1'),
              onFieldSubmitted: (_) => dropdownFocus.requestFocus(),
              textInputAction: TextInputAction.next,
            ),
            DropdownButtonFormField(
              focusNode: dropdownFocus,
              decoration: InputDecoration(
                labelText: 'Test Field 2',
              ),
              value: dropdownValue,
              items: [
                DropdownMenuItem(
                  value: 'a',
                  child: Text('A'),
                ),
                DropdownMenuItem(
                  value: 'b',
                  child: Text('B'),
                )
              ],
              onChanged: (newValue) {
                setState(() {
                  dropdownValue = newValue;
                });
              },
            ),
          ],
        ),
      ),
    );
  }
}
`

**Comment by GabrieleSuzart** [comment]
Hi, I'm having the same problem! I put a DropdownButton on the screen and then tested it with VoiceOver screen reader, after selecting a new option the reader focus is directed to the top of the screen, reading everything again, which makes the experience really bad for blind people.
